File: add_scale_bar.R
library(magick)
library(tiff)

add_scale_bar <- function(original_file_name, image_path, csv_path, x_coord, y_coord) {
  # Read the scale information from the CSV file
  scale_info <- read.csv(csv_path)
  
  # Determine the scale size and label based on camera make and match pattern
  scale_size_pixels <- NULL
  scale_label <- NULL
  for (row in 1:nrow(scale_info)) {
    if (grepl(scale_info$match_pattern[row], original_file_name)) {
      scale_size_pixels <- as.integer(scale_info$scale_size_pixels[row])
      scale_label <- as.character(scale_info$scale_label[row])
      break
    }
  }
  
  if (is.null(scale_size_pixels)) {
    print("No matching scale information found.")
    return(NULL)
  }
  
  # Read the original image
  original_image <- image_read(image_path)
  
  # Get the dimensions of the original image
  width <- image_info(original_image)$width[1]
  height <- image_info(original_image)$height[1]
  
  # Adjust y_coord for bottom-left origin
  y_coord <- height - y_coord
  
  # Create a scale bar
  scale_bar <- image_blank(width = width, height = height, color = "none") %>%
    image_draw() 
  rect(xleft = x_coord, ybottom = y_coord - 10, xright = x_coord + scale_size_pixels, ytop = y_coord, 
       col = "white", border = "white")
  dev.off()
  
  # Create a label as a separate image using base R graphics
  label_file <- tempfile(fileext = ".png")
  png(label_file, width = width, height = height, bg = "transparent")
  par(mar = c(0, 0, 0, 0))
  plot(0, 0, type = "n", xlim = c(0, width), ylim = c(0, height), ann = FALSE, axes = FALSE)
  text(x = x_coord-100, y = height - y_coord-50, labels = scale_label, col = "white", cex = 6, pos=1, offset=)  # Adjust 'cex' for size
  dev.off()
  
  # Read the label image
  label_text_image <- image_read(label_file)
  
  # Merge and flatten images
  final_combined_image <- image_flatten(c(original_image,label_text_image,scale_bar), operator = "over")
  
  # Create 'output' directory if it doesn't exist
  output_dir <- "images_with_scale"
  if (!dir.exists(output_dir)) {
    dir.create(output_dir)
  }
  
  # Save the final combined image in 'output' directory
  output_file_path <- file.path(output_dir, basename(sub("\\.[^.]*$", "_scaled.tif", original_file_name)))
  image_write(final_combined_image, output_file_path)
  
  # Optionally, delete the label file to clean up
  if (file.exists(label_file)) {
    file.remove(label_file)
  }
}

File: add_scale_bar.py

import pandas as pd
from PIL import Image, ImageDraw, ImageFont
import os

# Function to read and process the scale_info.csv file
def read_scale_info(csv_path):
    scale_info = pd.read_csv(csv_path)
    return scale_info

# Function to add a scale bar to an image
def add_scale_bar(image_path, scale_info, x_coord, y_coord, output_dir):
    original_file_name=os.path.basename(image_path)

    # Determine the scale size and label based on camera make and match pattern
    scale_row = scale_info[scale_info['match_pattern'].apply(lambda x: x in original_file_name)]
    if scale_row.empty:
        print("No matching scale information found.")
        return None

    scale_size_pixels = int(scale_row.iloc[0]['scale_size_pixels'])
    scale_label = scale_row.iloc[0]['scale_label']

    # Read the original image
    with Image.open(image_path) as img:
        draw = ImageDraw.Draw(img)
        
        # Adjust y_coord for bottom-left origin
        width, height = img.size
        y_coord = height - y_coord
        
        # Draw the scale bar
        draw.rectangle([(x_coord, y_coord), (x_coord + scale_size_pixels, y_coord+10)], fill="white")
        
        # Add scale label
        try:
            font = ImageFont.truetype("Arial Bold.ttf", 60)  # Using Arial font
        except IOError:
            font = ImageFont.load_default()
        draw.text((x_coord, y_coord - 65), scale_label, fill="white", font=font)
        
        # Save the image
        if not os.path.exists(output_dir):
            os.makedirs(output_dir)
        output_file_path = os.path.join(output_dir, os.path.splitext(os.path.basename(image_path))[0] + "_scaled.tif")
        img.save(output_file_path)

    return output_file_path

# Sample usage
csv_path = 'scale_info.csv'  # Update with the actual path of the CSV file
scale_info = read_scale_info(csv_path)
image_path = '4429479_Oxycorynus_melanocerus_h_d_CF3_P1_HF23.tif'
output_dir = "images_with_scale"

add_scale_bar(image_path, scale_info,  100, 100, output_dir) 
